collect_app/src/test/java/org/odk/collect/android/external/FormUriActivityTest.kt

```diff
@@ -0,0 +1,158 @@
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.intent.Intents
+import androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent
+import androidx.test.espresso.intent.matcher.IntentMatchers.hasData
+import androidx.test.espresso.intent.matcher.IntentMatchers.hasExtra
+import androidx.test.espresso.matcher.RootMatchers.isDialog
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
+import androidx.test.espresso.matcher.ViewMatchers.withId
+import androidx.test.espresso.matcher.ViewMatchers.withText
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.gson.Gson
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.`is`
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.kotlin.mock
+import org.mockito.kotlin.whenever
+import org.odk.collect.android.R
+import org.odk.collect.android.activities.FormEntryActivity
+import org.odk.collect.android.injection.config.AppDependencyModule
+import org.odk.collect.android.preferences.source.SettingsProvider
+import org.odk.collect.android.projects.CurrentProjectProvider
+import org.odk.collect.android.support.CollectHelpers
+import org.odk.collect.projects.InMemProjectsRepository
+import org.odk.collect.projects.Project
+import org.odk.collect.projects.ProjectsRepository
+import org.odk.collect.shared.strings.UUIDGenerator
+
+@RunWith(AndroidJUnit4::class)
+class FormUriActivityTest {
+    private lateinit var projectsRepository: ProjectsRepository
+    private lateinit var currentProjectProvider: CurrentProjectProvider
+    private val context = ApplicationProvider.getApplicationContext<Application>()
+    private val firstProject = Project.DEMO_PROJECT
+    private val secondProject = Project.Saved("123", "Second project", "S", "#cccccc")
+
+    @Before
+    fun setup() {
+        projectsRepository = InMemProjectsRepository()
+        currentProjectProvider = mock()
+
+        projectsRepository.save(firstProject)
+        projectsRepository.save(secondProject)
+        whenever(currentProjectProvider.getCurrentProject()).thenReturn(firstProject)
+
+        CollectHelpers.overrideAppDependencyModule(object : AppDependencyModule() {
+            override fun providesProjectsRepository(
+                uuidGenerator: UUIDGenerator?,
+                gson: Gson?,
+                settingsProvider: SettingsProvider?
+            ): ProjectsRepository {
+                return projectsRepository
+            }
+
+            override fun providesCurrentProjectProvider(
+                settingsProvider: SettingsProvider?,
+                projectsRepository: ProjectsRepository?
+            ): CurrentProjectProvider {
+                return currentProjectProvider
+            }
+        })
+    }
+
+    @Test
+    fun `When there are no projects then display alert dialog`() {
+        projectsRepository.deleteAll()
```

Comment:
Personally I'd find these tests easier to read if the projects were added at the start of the test rather than in `setup`. Then, the state of the world (the "given") can be seen next to the actions and assertions (the "when" and "then").

Reply:
Done.